Add support for powernv builds to qemu tests

nick-child-ibm opened the pull request:
Hello,

I am not sure if this is worth a pull request or not but I figured it helps me so maybe it can be beneficial upstream.

I wanted to try to build and test a kernel configured for powernv. I had built my kernel with the recommended command `sudo make SRC=../../linux-git/ kernel@ppc64le@ubuntu@21.04 DEFCONFIG=powernv_defconfig JFACTOR=$(nproc)` . But the boot script, `qemu-pseries` kept failing. More specifically, it would hang indefinitely when it tried to boot from memory. I was able to resolve the issue by using `powernv` for the `-machine` qemu parameter (previously set to `pseries,cap-htm=off`). I added the ability for the user to set a custom machine type and made a `qemu-powernv` script that calls `qemu-pseries` with the appropriate environment variables set.

Anyway, I would be interested to know if I am totally wrong with this approach. There are too many qemu options to keep track of and its hard to find the best solution.
Thanks! 

Side question: I should probably open an issue for this but I have a feeling it is a misunderstanding on my end. I was having issues with booting the default `ppc64le_defconfig` build. The script `qemu-pseries` kept failing when it got to the network setup function. It could not find the eth0 network device. I was able to fix this issue by setting the network interface card to `e1000` instead of `virtio-net-pci`. The reason I am more confused is because I thought that virtio was supposed to recognize the host hardware and virtualize based on that. It seems weird that the Intel specific driver would work instead of the intuitive driver. That being said I am using an Intel NIC on my host so maybe that lowers the amount of things to go wrong in virtualization land. Anyway, apologies for the long side question, but I will suck it up if it is not an issue for anyone else. Otherwise, I can add a commit for user specified `-nic` parameters.

mpe commented:
Hey, sorry I only just saw this. Not sure why but I didn't get an email about it.

Congrats on wading through the mess far enough to identify that the qemu machine type needs changing :sweat_smile: 

There's a few differences between the pseries and powernv machines, there are different powernv machine types per CPU type, and also the machine type in `/proc/cpuinfo` should differ (see the check on line 71).

I think what we want to do is have a separate script for pseries and powernv. They will be similar, but I think that's OK.

So for starters just copy the `qemu-pseries` script to `qemu-powernv`. Then for powernv you need to set the machine type to match the CPU, so eg. for cpu=POWER8 you need machine=powernv8, and for cpu=POWER9 machine=powernv9.

And the output of `/proc/cpuinfo` should contain `IBM PowerNV (emulated by qemu)`.

For your second question I think you just need to use `ppc64le_guest_defconfig`, that has the guest drivers built-in, including the virtio drivers.

nick-child-ibm committed:
Add support for powernv builds to qemu tests

A kernel built with the powernv defconfig should be tested on an
emulated powernv machine. To do this with qemu, the `-machine`
parameter must be given a powernv value that aligns with the cpu type.
For now the only two options are `powernv8` and `powernv9` for `power8`
and `power9` respectively. Other test conditions must also be adjusted,
the machine type should look for powernv insted of pseries. To address
these differences, a `qemu-powernv` script has been created, it closely
resembles `qemu-pseries` but makes the appropriate adjustments for the
machine differences.

nick-child-ibm commented:
Hey Michael, thanks for getting back to me!
Now it is my turn to apologize for not responding XD . I have been busy with a couple other measures and could not get around to this until late last night. 

Thanks for the recommendations, I have made the adjustments to the commit. Also, your advice for using `ppc64le_guest_defconfig` worked marvelously so thanks for that. 

The new commit "successfully" tests a powernv kernel. Kinda...
It works fine until the last step where `halt` is run and EOF is expected. For some odd reason that I could not figure out, powernv qemu instances do not poweroff properly. Or maybe it does and it just does not exit the emulation back to the host terminal nicely like pseries. If this is the case then disregard the below statements and I can edit the script to expect something other than EOF after the `halt` command is run. 

I did some investigation and found that the "issue" may be in skiboot. When `halt` is called: pseries will call `rtas_halt` and everything works nicely, powernv will call `pnv_halt` which will eventually hang in a loop in `pnv_power_off` (this power off process should work as it is documented by skiboot [here](https://github.com/open-power/skiboot/blob/master/doc/opal-api/opal-cec-power-down-5.rst)). `pnv_power_off` finds its way to skiboot land via `opal_cec_power_down` which then, for powernv on qemu, calls `astbmc_ipmi_power_down` -> `ipmi_chassis_control` . This is where I am once again confused about whether something is wrong or not,  `ipmi_chassis_control` will return early with OPAL_CLOSED since `ipmi_present()` returns false. This return code is returned all the way back to `pnv_power_off` but the documentation (linked above) does not describe the case when this function returns OPAL_CLOSED. So, my concerns are that maybe this return code is not expected and we should take a further action when this happens. I am really not sure if this behavior is intentional or not, the main thing is it seems strange that there is no way to shutdown the powernv instance and return back the command prompt.

mpe commented:
Hi Nick, sorry for yet another long delay. I've finally pushed a whole bunch of changes I had been using internally, including a powernv script. In the end I did share the script between pseries and powernv, as you originally proposed, sorry about that. Let me know if what I pushed works for you. I'll close this for now, feel free to reopen if you want.
